### PR TITLE
Specialize http2.StreamError when the server reads frames

### PR DESCRIPTION
**Pull request by iamqizhao**
updates #597

### CODE REVIEW

File: test/end2end_test.go [hunk 1]
```diff
@@ -80,6 +80,9 @@ var (
 		"tkey1": []string{"trailerValue1"},
 		"tkey2": []string{"trailerValue2"},
 	}
+	malformedHTTP2Metadata = metadata.MD{
+		"Key": []string{"foo"},
```
Comment (bradfitz) [inline test/end2end_test.go]: I might add a comment to this line saying // capital "Key" is illegal in HTTP/2

File: test/end2end_test.go [hunk 2]
```diff
@@ -889,6 +892,35 @@ func testMetadataUnaryRPC(t *testing.T, e env) {
 	}
 }
 
+func TestMalformedHTTP2Metadata(t *testing.T) {
```
Comment (bradfitz) [inline test/end2end_test.go]: Add some prose before this, explaining at a high level what this test is testing.

**Comment by bradfitz** [comment]
LGTM. Left some comments.

**Comment by iamqizhao** [comment]
Thanks. All done.

**Pull request merged**